kong/runloop/handler.lua

```diff
@@ -1318,33 +1322,22 @@ return {
 
       var.upstream_scheme = balancer_data.scheme
 
-      do
-        -- set the upstream host header if not `preserve_host`
-        local upstream_host = var.upstream_host
-        local upstream_scheme = var.upstream_scheme
-
-        if not upstream_host or upstream_host == "" then
-          upstream_host = balancer_data.hostname
-
-          if upstream_scheme == "http"  and balancer_data.port ~= 80 or
-             upstream_scheme == "https" and balancer_data.port ~= 443 or
-             upstream_scheme == "grpc"  and balancer_data.port ~= 80 or
-             upstream_scheme == "grpcs" and balancer_data.port ~= 443
-          then
-            upstream_host = upstream_host .. ":" .. balancer_data.port
-          end
+      local ok, err = balancer.set_host_header(balancer_data)
+      if not ok then
+        ngx.log(ngx.ERR, "failed to set balancer Host header: ", err)
 
-          var.upstream_host = upstream_host
-        end
+        return ngx.exit(500)
```

Comment:
The most probable cause of failure here is `ngx_create_request` not being able to allocate buffers. So, this is a temporary condition, and we know what happened. Therefore, a 503 seems more appropriate to me.

Reply:
I think either one is fine here. We have traditionally used 500 to indicate any kind of error Kong experiences during request processing, so it should be okay to use 500 here as well.

It may be difficult to determine if the failure reason is temporary or not here (memory leaks, for example, could last for a long time until being manually corrected). So I'm okay with keeping 500 here.